### PR TITLE
fix: PureCN rules

### PR DESCRIPTION
**Pull request by khurrammaqbool**
## Description

PureCN rule fails due to bash strict mode when purity and ploidy can not be estimated. This causes the whole pipeline to terminate prematurely.

#### Fixed

- PureCN rules to bypass bash strict mode in both TO and TN workflows.

## Documentation

- [x] N/A
- [ ] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - [Document Name]

### Feature Tests

- [x] N/A
- [x] Test [Description]
    - [Screenshot]

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### User Changes

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [ ] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [ ] Verified and updated documentation if necessary.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [x] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [x] Code implements the intended features or fixes the reported issue.
    - [x] Code follows the project's coding standards and style guide.
- **Documentation**
    - [x] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [x] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [x] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -1,4 +1,11 @@
-[14.0.0]
+[[14.0.1]
```
Comment (mathiasbio) [inline CHANGELOG.rst:1]:
```suggestion
[14.0.1]
```

**Review by mathiasbio — COMMENTED** [review]
Shouldn't the merge branch be the release branch, not master? Or maybe you and @ivadym have come to some clever solution 🌟

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule [hunk 2]
```diff
@@ -200,29 +205,32 @@ Rscript $PURECN \
 --fun-segmentation Hclust \
 --force --post-optimize \
 --seed 124;
+}} || exit_status="false"
 
-if [[ -f "{params.tmpdir}/{params.sample_id}_dnacopy.seg" ]];  then
-mv {params.tmpdir}/{params.sample_id}_dnacopy.seg {params.cnv_csv};
-fi;
+if $exit_status; then
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:210]:
```suggestion
if [ "$exit_status" = "true" ]; then
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_only_tga.rule [hunk 3]
```diff
@@ -138,29 +142,32 @@ Rscript $PURECN \
 --fun-segmentation Hclust \
 --force --post-optimize \
 --seed 124;
+}} || exit_status="false"
 
-if [[ -f "{params.tmpdir}/{params.sample_id}_dnacopy.seg" ]];  then
-mv {params.tmpdir}/{params.sample_id}_dnacopy.seg {params.cnv_csv};
-fi;
+if $exit_status; then
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_only_tga.rule:147]:
```suggestion
if [ "$exit_status" = "true" ]; then
```

File: CHANGELOG.rst (hunk 1 above)
Comment (ivadym) [inline CHANGELOG.rst:1]: And don't forget to change the base branch to release_v14.0.1 😄 

The documentation mathias updated yesterday if you want to have a look: https://atlas.scilifelab.se/infrastructure/BALSAMIC/balsamic_gitmodel/#steps-for-making-a-new-release

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule [hunk 4]
```diff
@@ -200,29 +205,32 @@ Rscript $PURECN \
...
+    if [[ -f "{params.tmpdir}/{params.sample_id}_genes.csv" ]];  then
+    mv {params.tmpdir}/{params.sample_id}_genes.csv {params.loh_genes};
+    fi;
 
-if [[ -f "{params.tmpdir}/{params.sample_id}.vcf.gz" ]];  then
-mv {params.tmpdir}/{params.sample_id}.vcf.gz {params.purecn_vcf};
-fi;
+    if [[ -f "{params.tmpdir}/{params.sample_id}.vcf.gz" ]];  then
+    mv {params.tmpdir}/{params.sample_id}.vcf.gz {params.purecn_vcf};
+    fi;
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:229]: If you add indentation  here, let's also do it for the else ⭐

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_only_tga.rule [hunk 5]
```diff
@@ -138,29 +142,32 @@ Rscript $PURECN \
...
+    if [[ -f "{params.tmpdir}/{params.sample_id}.vcf.gz" ]];  then
+    mv {params.tmpdir}/{params.sample_id}.vcf.gz {params.purecn_vcf};
+    fi;
 
-if [[ -f "{params.tmpdir}/{params.sample_id}.csv" ]];  then
-mv {params.tmpdir}/{params.sample_id}.csv {output.purecn_purity};
+    if [[ -f "{params.tmpdir}/{params.sample_id}.csv" ]];  then
+    mv {params.tmpdir}/{params.sample_id}.csv {output.purecn_purity};
+    fi;
 else
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_only_tga.rule:171]: same comment here for the indentation

File: CHANGELOG.rst [hunk 6]
```diff
@@ -1,4 +1,11 @@
...
+-------
+
+Fixed:
+^^^^^^
+* PureCN fail due to bash strict mode ASCAT-Ngs container https://github.com/Clinical-Genomics/BALSAMIC/pull/
+
+14.0.0]
```
Comment (ivadym) [inline CHANGELOG.rst:8]:
```suggestion
[14.0.0]
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule (hunk 4 above)
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:229]: Let us not touch this part as it may add this indentation in the output.

**Comment by khurrammaqbool** [comment]
Screenshot 2024-02-29 at 13 07 15

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule [hunk 7]
```diff
@@ -200,29 +205,32 @@ Rscript $PURECN \
...
 
-if [[ -f "{params.tmpdir}/{params.sample_id}.csv" ]];  then
-mv {params.tmpdir}/{params.sample_id}.csv {output.purecn_purity};
+    if [[ -f "{params.tmpdir}/{params.sample_id}.csv" ]];  then
+    mv {params.tmpdir}/{params.sample_id}.csv {output.purecn_purity};
+    fi;
 else
 echo '"Sampleid","Purity","Ploidy","Sex","Contamination","Flagged","Failed","Curated","Comment"
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:235]: Could you just update the comment as well since we are touching this part to say that the default values of purity/ploidy has been used if purecn fails 🙏 🙏

**Review by ivadym — APPROVED** [review]
Well done 💯

**Review by ivadym — APPROVED** [review]
Nice ⭐
